### ISSUE TITLE
Print location of the dead links

### ISSUE THREAD

**Issue by DarkEld3r** [deadlinks/cargo-deadlinks#14]
Sometimes it isn't trivial to find where dead link is actually located. I have to modify `cargo-deadlinks` locally with additional `debug!` print in order to find it. :sweat_smile: It would be nice to provide additional information.

**Comment by hobofan** [comment]
Agree, would be a great future to have!

However with the current approach of analyzing the HTML output it's not really trivial to introduce.

AFAIK the solution with the best UX would very likely require nightly and involve a compiler plugin.

------

##### Compiler plugin variant (requires nightly)

- Inject some kind of marker into every link (e.g. a UUID anchor tag) via a compiler plugin
- The compiler plugin can then build a lookup table of UUID->position. The position can be determined via [rust-docstrings](https://github.com/killercup/rust-docstrings); see also [ronat](https://github.com/hobofan/ronat) where I've done something similar
- When we find a broken link we can then get the code location from the lookup table

##### Stable variant - requires manual intervention and only works with comment-block granularity

- Manually place a marker into the documentation like `<!--- deadlinks-id: 123 --->`
- Since those comments stay in the generated output we can then find the appropriate comment block via `ripgrep` or similar

**Comment by hobofan** [comment]
Duplicate of #3, but this one stays since the other one has old information and I've written more here 😅

**Comment by hobofan** [comment]
@DarkEld3r Having to modify `cargo-deadlinks` also shouldn't be necessary right now. What exactly are you logging that helps you pinpoint the issue?

Have you tried using the `--debug` flag? That should usually give you a rough starting point.

**Comment by DarkEld3r** [comment]
Now I see, that I missed (`info!`) logging in the `parse_html_file` function. :sweat_smile: That will be enough for me.

**Comment by hobofan** [comment]
https://github.com/rust-lang/rust/pull/47046 recently landed which should combat the problem `cargo-deadlinks` was built for to some degree.

If we were to have this feature here implemented, we could probably also lint for the style in which a link was specified, and point people towards the new syntax of specifying links (maybe this should also be done in rustdoc directly).

**Comment by Marwes** [comment]
Just printing the html file where the dead link was found would be nice. That would narrow the search down a lot and is useful for non-rustdoc .html as well.

**Comment by jyn514** [comment]
> Just printing the html file where the dead link was found would be nice. That would narrow the search down a lot and is useful for non-rustdoc .html as well.

The output now looks like this:
```
checking directory "/home/joshua/.local/lib/cargo/target/doc/toggle"
Found invalid urls in /home/joshua/.local/lib/cargo/target/doc/toggle/struct.Args.html:
	Linked file at path /home/joshua/.local/lib/cargo/target/doc/structopt/trait.StructOptInternal.html does not exist!
	Linked file at path /home/joshua/.local/lib/cargo/target/doc/structopt/trait.StructOptInternal.html does not exist!
	Linked file at path /home/joshua/.local/lib/cargo/target/doc/structopt/trait.StructOptInternal.html does not exist!
	Linked file at path /home/joshua/.local/lib/cargo/target/doc/structopt/trait.StructOptInternal.html does not exist!
```

Is that clear enough to find what went wrong, or would you like it to be more specific? Tracing that back to the source file will be very hard like hobofan said, but adding a line number should be pretty easy.

**Comment by Marwes** [comment]
Yeah, as long as I have the link it is easy to find where it comes from (though having the line directly ould of course be better)
